cljfmt/src/cljfmt/tool.clj

```diff
@@ -99,25 +100,77 @@
   ([a]   a)
   ([a b] (merge-with + a b)))
 
-(defn check-no-config
-  "The same as `check`, but ignores dotfile configuration."
-  [options]
-  (let [map*   (if (:parallel? options) pmap map)
-        counts (->> (:paths options)
-                    (mapcat (partial find-files options))
-                    (map* (partial check-one options))
-                    (map (fn [status]
-                           (print-file-status options status)
-                           (:counts status)))
-                    (reduce merge-counts))]
+(defn- merge-statuses
+  [a b]
+  (let [^File file (:file b)
+        path       (.getPath file)
+        diff       (:diff b)
+        exception  (:exception b)]
+    (cond-> a
+      true (update :counts #(merge-with + % (:counts b)))
+
+      (and file (= 1 (-> b :counts :incorrect)))
+      (assoc-in [:incorrect path] "has incorrect formatting")
+
+      (and file diff) (assoc-in [:incorrect path] diff)
+      (and file exception) (assoc-in [:error path] exception))))
+
+(defmulti report-cli (fn [event & _] event))
+(defmulti report-lib (fn [event & _] event))
+
+(defmethod report-cli :check/initial
+  [_ _ _]
+  nil)
+
+(defmethod report-cli :check/file
+  [_ context data]
+  (print-file-status context data)
+  (:counts data))
+
+(defmethod report-cli :check/summary
+  [_ summary _]
+  (let [counts (reduce merge-counts summary)]
     (print-final-count counts)
     (exit counts)))
 
+(defmethod report-lib :check/initial
+  [_ _ _]
+  nil)
+
+(defmethod report-lib :check/file
+  [_ _ data]
+  data)
+
+(defmethod report-lib :check/summary
+  [_ context _]
+  (reduce merge-statuses {} context))
+
+(defn check-no-config
+  "The same as `check`, but ignores dotfile configuration."
+  [{:keys [parallel? paths report] :as options
+    :or   {report report-cli}}]
+  (report :check/initial options nil)
+  (let [map*    (if parallel? pmap map)
+        summary (->> paths
+                     (mapcat (partial find-files options))
+                     (map* (partial check-one options))
+                     (map #(report :check/file options %)))]
```

Comment:
The `report` should be a `reduce`, not a map. The reporting functions should be a type signature of:

```clojure
(report event context data) -> new-context
```

That way the type signature of the multi-method remains consistent. See the example in my earlier comment.

Reply:
Yeah, I started with that. It's a little more complex and messy that way. Instead of just returning what is inherently a list of results, you have to stuff it into the original context somewhere, which starts out as the `options` map. So then the final return value is all of those options + wherever you stuffed the results.

I have a working refactor that puts them into a `:results` key. Happy to push that if you still prefer it.